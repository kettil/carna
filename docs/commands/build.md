# The `build` task

Build the application

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
