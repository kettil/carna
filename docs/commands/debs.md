# The `debs` task

Checks if there are orphaned packages

## Options

| Options | Description       |
| ------- | ----------------- |
| --ci    | Run it in CI mode |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
