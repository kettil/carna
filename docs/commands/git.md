# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
