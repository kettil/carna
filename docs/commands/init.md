# The `init` task

Creates a new project with all config files in the current folder.

## Options

| Options    | short | Description                                                  |
| ---------- | ----- | ------------------------------------------------------------ |
| --package  | -p    | Project is created as a package                              |
| --cli      | -c    | Extends the package with CLI features (required `--package`) |
| --noCommit |       | No initial commit is executed at the end                     |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
