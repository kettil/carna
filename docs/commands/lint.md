# The `lint` task

Run the code quality tools

## Options

| Options | Description                                      |
| ------- | ------------------------------------------------ |
| --ci    | Run it in CI mode                                |
| --only  | Run a single code quality tool (required `--ci`) |

## Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
