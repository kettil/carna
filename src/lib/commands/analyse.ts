import { createBuilder, createHandler } from '../cli/yargs';
import type { AnalyseProps } from '../tasks/analyseTask';
import { analyseServices, analyseTask } from '../tasks/analyseTask';

const command = 'analyse';
const desc = 'Run code quality tools';

const aliases = ['$0'];
const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<AnalyseProps>(analyseTask);
const builder = createBuilder<AnalyseProps>(command, (yargs) =>
  yargs.options({
    all: {
      ...boolOptions,
      alias: 'a',
      describe: 'All files are checked and not only the changed/new files',
    },
    only: {
      ...options,
      alias: 'o',
      choices: analyseServices,
      describe: 'Run a single code quality tool',
    },
  }),
);

export { aliases, command, desc, builder, handler };
