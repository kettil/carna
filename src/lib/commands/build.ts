import { createBuilder, createHandler } from '../cli/yargs';
import { buildTask, BuildProps } from '../tasks/buildTask';

const command = 'build';
const desc = 'Run the build process';

const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<BuildProps>(buildTask);
const builder = createBuilder<BuildProps>(command, (yargs) =>
  yargs.options({
    watch: {
      ...boolOptions,
      alias: 'w',
      describe: 'Watch files for changes and rebuild the changed files (without type files)',
    },
  }),
);

export { command, desc, builder, handler };
