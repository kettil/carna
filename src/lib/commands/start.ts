import { createBuilder, createHandler } from '../cli/yargs';
import type { StartProps } from '../tasks/startTask';
import { startTask } from '../tasks/startTask';

const command = 'start';
const desc = 'Run application directly';

const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean' } as const;

const handler = createHandler<StartProps>(startTask);
const builder = createBuilder<StartProps>(command, (yargs) =>
  yargs.options({
    watch: {
      ...boolOptions,
      alias: 'w',
      describe: 'Reload your app on source file changes',
    },
    script: {
      ...options,
      type: 'string',
      alias: 's',
      describe: 'Script to be called (default: src/index.[ts|js])',
    },
    // eslint-disable-next-line @typescript-eslint/naming-convention -- terminal name schema
    'build-dependencies': {
      ...boolOptions,
      alias: 'b',
      describe: 'In a monorepo, all packages are built beforehand',
    },
    // eslint-disable-next-line @typescript-eslint/naming-convention -- terminal name schema
    'clear-console': {
      ...boolOptions,
      alias: 'c',
      implies: ['watch'],
      conflicts: ['pino-pretty'],
      describe: 'Clear console on each restart (only with watch mode)',
    },
    // eslint-disable-next-line @typescript-eslint/naming-convention -- terminal name schema
    'pino-pretty': {
      ...boolOptions,
      alias: 'p',
      conflicts: ['clear-console'],
      describe: 'Enables the pino log formatter (pino-pretty)',
    },
  }),
);

export { command, desc, builder, handler };
