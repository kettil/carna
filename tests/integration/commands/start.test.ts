import * as fs from 'fs';
import type { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/start';
import type { PropsGlobal } from '../../../src/lib/types';
import { getArgv } from '../../shared/setup/argv';
import { mockFilePackage, packageJsonWithWorkspaces } from '../../shared/setup/packageJson';
import { getReadFileWithHooksFiles, getReadFileFiles } from '../../shared/setup/readFileFiles';

describe('command start', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    expect.assertions(2);

    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  test('it should work', async () => {
    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test.each([
    [false, false],
    [false, true],
    [true, false],
    [true, true],
  ])('it should work with workspaces [watch: %p, buildDependencies: %p]', async (watch, buildDependencies) => {
    (fs as any).setMockReadFileFiles(
      getReadFileFiles({
        [mockFilePackage]: JSON.stringify(packageJsonWithWorkspaces),
      }),
    );

    const result = await handler(getArgv({ watch, buildDependencies, script: 'packages/a/src/index.ts' }));

    expect(result).toBeUndefined();
  });
});
