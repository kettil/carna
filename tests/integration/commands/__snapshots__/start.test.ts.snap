// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command start it should be create the expected builder: options 1`] = `
Array [
  Object {
    "build-dependencies": Object {
      "alias": "b",
      "default": false,
      "describe": "In a monorepo, all packages are built beforehand",
      "group": "start-Options",
      "type": "boolean",
    },
    "clear-console": Object {
      "alias": "c",
      "default": false,
      "describe": "Clear console on each restart (only with watch mode)",
      "group": "start-Options",
      "type": "boolean",
    },
    "script": Object {
      "alias": "s",
      "describe": "Script to be called (default: src/index.[ts|js])",
      "group": "start-Options",
      "type": "string",
    },
    "watch": Object {
      "alias": "w",
      "default": false,
      "describe": "Reload your app on source file changes",
      "group": "start-Options",
      "type": "boolean",
    },
  },
]
`;

exports[`command start it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 start [options]",
]
`;

exports[`command start it should work with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:start",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command start it should work with hooks: child-process 2`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "/path/to/project/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-node",
  "cwd": "/path/to/project",
  "env": Object {
    "DOTENV_CONFIG_PATH": ".env",
  },
}
`;

exports[`command start it should work with hooks: child-process 3`] = `
Object {
  "args": Array [
    "run",
    "postcarna:start",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command start it should work with hooks: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work with hooks: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work with hooks: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: false]: child-process 1`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "/path/to/project/packages/a/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-node",
  "cwd": "/path/to/project/packages/a",
  "env": Object {
    "DOTENV_CONFIG_PATH": "../../.env",
  },
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: false]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: false]: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: false]: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: false]: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: child-process 1`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: child-process 2`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "/path/to/project/packages/a/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-node",
  "cwd": "/path/to/project/packages/a",
  "env": Object {
    "DOTENV_CONFIG_PATH": "../../.env",
  },
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work with workspaces [watch: false, buildDependencies: true]: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: false]: child-process 1`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--watch",
    "/path/to/project/packages/b/build",
    "/path/to/project/packages/a/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-watch",
  "cwd": "/path/to/project/packages/a",
  "env": Object {
    "DOTENV_CONFIG_PATH": "../../.env",
  },
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: false]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: false]: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: false]: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: false]: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: child-process 1`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: child-process 2`] = `
Object {
  "args": Array [
    "-d",
    "build",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--watch",
    "--skip-initial-build",
    "src",
  ],
  "command": "/path/to/project/node_modules/.bin/babel",
  "cwd": "/path/to/project/packages/b",
  "env": Object {},
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: child-process 3`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "--watch",
    "/path/to/project/packages/b/build",
    "/path/to/project/packages/a/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-watch",
  "cwd": "/path/to/project/packages/a",
  "env": Object {
    "DOTENV_CONFIG_PATH": "../../.env",
  },
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: glob 1`] = `
Object {
  "options": Object {
    "absolute": true,
    "cwd": "/path/to/project",
  },
  "pattern": "packages/*",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;

exports[`command start it should work with workspaces [watch: true, buildDependencies: true]: spawn 1`] = `2`;

exports[`command start it should work: child-process 1`] = `
Object {
  "args": Array [
    "--extensions",
    ".js,.jsx,.ts,.tsx,.json",
    "--config-file",
    "/path/to/project/babel.config.js",
    "/path/to/project/src/index.ts",
  ],
  "command": "/path/to/project/node_modules/.bin/babel-node",
  "cwd": "/path/to/project",
  "env": Object {
    "DOTENV_CONFIG_PATH": ".env",
  },
}
`;

exports[`command start it should work: process-stdin 1`] = `
Object {
  "mode": true,
}
`;

exports[`command start it should work: process-stdin 2`] = `
Object {
  "encoding": "utf8",
}
`;

exports[`command start it should work: process-stdin 3`] = `
Object {
  "callback": [Function],
  "event": "data",
}
`;
